### ISSUE TITLE
Trying to evaluate during training produces KeyError `eval_loss`

### ISSUE THREAD

**Issue by KathyHaem** [urchade/GLiNER#140]
Hi there; I tried to use the train.py example but added `load_best_model_at_end` as a parameter. What that does is the Trainer will try to run inference on an evaluation set every X steps (I'm passing in an evaluation set in the same format as the training set).
However, it doesn't compute any real metric on this evaluation set, not even a loss (hence the KeyError `eval_loss`), and then doesn't have anything to decide the best model on.

The last line here is where it ends up throwing (in `_maybe_log_save_evaluate` of the Trainer class)
```
        metrics = None
        if self.control.should_evaluate:
            metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
            self._report_to_hp_search(trial, self.state.global_step, metrics)

            # Run delayed LR scheduler now that metrics are populated
            if isinstance(self.lr_scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
                metric_to_check = self.args.metric_for_best_model
                if not metric_to_check.startswith("eval_"):
                    metric_to_check = f"eval_{metric_to_check}"
                self.lr_scheduler.step(metrics[metric_to_check])
```
Getting this to work would be crucial for better fine-tuning. Right now I can't even check if or when I'm overfitting on the training set. How could this work? It seems pretty clear to me that the parent class Trainer doesn't know what to do with the GliNER eval dataset.

**Comment by andrea-mar** [comment]
@KathyHaem i had the same issue. I have started a thread in Discussions. The discord channel was helpful: https://discord.com/channels/1227934940457336923/1228426388093014168/1255580142533869588
You will need to add a wrapper function to the Trainer  - the code is in the link above.

**Comment by urchade** [comment]
@Ingvarstep may help in this

**Comment by Ingvarstep** [comment]
@KathyHaem, yes, you are right. The Trainer class wasn't able to handle evaluation correctly. I resolved this issue by adding a custom prediction_step function within the Trainer class. 
https://github.com/urchade/GLiNER/blob/c07ebe4e24297820d496791e09cba275087ae16f/gliner/training/trainer.py#L169

**Comment by KathyHaem** [comment]
@Ingvarstep thanks for this, it can calculate an `eval_loss` now
